File: Core/Debugger.fs
namespace Ufrgs.Inf.ArchSims.Core

open System.Collections.Generic

module Debugger =

    type DebuggerStopReason =
    | None
    | Halted
    | Breakpoint
    | RunningForever

    type Debugger = {
        CpuGetProgramCounter: (unit -> int)
        CpuStep: (unit -> bool)                  // Must return 'true' when halted
        mutable InstructionCount: int
        mutable LastStop: DebuggerStopReason
        mutable Breakpoints: Set<int>
    }

    let CreateDebugger cpuGetProgramCounter cpuStep = {
        CpuGetProgramCounter = cpuGetProgramCounter
        CpuStep = cpuStep
        InstructionCount = 0
        LastStop = None
        Breakpoints = Set.empty
    }

    let DebuggerReset debugger =
        debugger.InstructionCount <- 0
        debugger.LastStop <- None
        debugger.Breakpoints <- Set.empty

    let DebuggerStep debugger =
        debugger.LastStop <- if debugger.CpuStep() 
                             then Halted
                             else if debugger.Breakpoints.Contains (debugger.CpuGetProgramCounter()) 
                                  then Breakpoint
                                  else None
        debugger.InstructionCount <- debugger.InstructionCount + 1

    let DebuggerRun debugger maximumInstructions =
        debugger.LastStop <- None
        let limitInstructions = debugger.InstructionCount + maximumInstructions
        while debugger.LastStop = None do
            DebuggerStep debugger
            if debugger.InstructionCount >= limitInstructions then
                debugger.LastStop <- RunningForever

    let DebuggerSetBreakpoint debugger address =
        debugger.Breakpoints <- debugger.Breakpoints.Add(address)

    let DebuggerClearBreakpoint debugger address =
        debugger.Breakpoints <- debugger.Breakpoints.Remove(address)

File: web/src/core/debugger.ts
export type DebuggerStopReason = 'none' | 'halted' | 'breakpoint' | 'runningForever';

export interface Debugger {
  cpuGetProgramCounter: () => number;
  cpuStep: () => boolean; // returns true when the CPU is halted
  instructionCount: number;
  lastStop: DebuggerStopReason;
  breakpoints: Set<number>;
}

export function createDebugger(
  cpuGetProgramCounter: () => number,
  cpuStep: () => boolean,
): Debugger {
  return {
    cpuGetProgramCounter,
    cpuStep,
    instructionCount: 0,
    lastStop: 'none',
    breakpoints: new Set(),
  };
}

export function debuggerReset(debug: Debugger): void {
  debug.instructionCount = 0;
  debug.lastStop = 'none';
  debug.breakpoints = new Set();
}

export function debuggerStep(debug: Debugger): void {
  if (debug.cpuStep()) {
    debug.lastStop = 'halted';
  } else if (debug.breakpoints.has(debug.cpuGetProgramCounter())) {
    debug.lastStop = 'breakpoint';
  } else {
    debug.lastStop = 'none';
  }
  debug.instructionCount++;
}

export function debuggerRun(debug: Debugger, maximumInstructions: number): void {
  debug.lastStop = 'none';
  const limit = debug.instructionCount + maximumInstructions;
  while (debug.lastStop === 'none') {
    debuggerStep(debug);
    if (debug.instructionCount >= limit) {
      debug.lastStop = 'runningForever';
    }
  }
}

export function debuggerSetBreakpoint(debug: Debugger, address: number): void {
  debug.breakpoints.add(address);
}

export function debuggerClearBreakpoint(debug: Debugger, address: number): void {
  debug.breakpoints.delete(address);
}
